### PR TITLE
np.max, np.min, np.nanmax, np.nanmin - support for complex dtypes

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit for CI.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Related https://github.com/numba/numba/issues/546
CC @seibert

**Comment by seibert** [comment]
With 4.5 years of hindsight, I think the answer is definitely "Do what NumPy does".  I don't think we did ourselves any favors in nearly situations when we tried to somehow split the difference between Python and NumPy.  :)

**Comment by rjenc29** [comment]
OK, the last commit should mean this is in reviewable state.

Let me know what you think when you get a mo, ta!

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -530,46 +581,83 @@ def _trivial_isnan(x):
...
-            raise ValueError("nanmin(): empty array")
-        for view in np.nditer(a):
-            minval = view.item()
-            break
-        for view in np.nditer(a):
-            v = view.item()
-            if not minval < v and not isnan(v):
-                minval = v
-        return minval
+@register_jitable
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Note to self. This is just the sort of place where `@inlinable` would be useful. This doesn't need to be compiled with reliance on the LLVM inliner to inline the function body, the IR just needs inlining at the call site.

File: numba/tests/test_array_reductions.py [hunk 2]
```diff
@@ -757,6 +757,79 @@ def test_ptp_exceptions(self):
...
+    def test_nanmin_nanmax_non_array_inputs(self):
+        pyfuncs = array_nanmin, array_nanmax
+
+        def check(a):
+            expected = pyfunc(a)
+            got = cfunc(a)
+            self.assertPreciseEqual(expected, got)
+
+        def a_variations():
+            yield [1, 6, 4, 2]
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py]: Perhaps add some `np.nan` values into some of the input here?

File: numba/targets/arraymath.py [hunk 3]
```diff
@@ -530,46 +573,83 @@ def _trivial_isnan(x):
...
+    if a.size == 0:
+        raise ValueError('zero-size array to reduction operation not possible')
+
+def nan_min_max_factory(comparison_op, is_complex_dtype):
+
+    if is_complex_dtype:
+        def impl(a):
+            arr = np.asarray(a)
+            check_array(arr)
+            for view in np.nditer(arr):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Is this another place where a `next(iterator).take(0)` might work? (and below)

**Pull request merged**